File: packages/server/app/models/session.py
"""Session and SessionEvent models for agent session tracking."""
from typing import Optional
from sqlalchemy import String, Text, Integer, BigInteger, ForeignKey, Index
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base


class Session(Base):
    """Agent session execution tracking."""
    __tablename__ = "sessions"
    __table_args__ = (
        Index("idx_sessions_agent_created", "agent_id", "created_at", postgresql_using="btree"),
    )
    
    id: Mapped[str] = mapped_column(String(128), primary_key=True)
    agent_id: Mapped[str] = mapped_column(String(128), nullable=False, index=True)
    source: Mapped[str] = mapped_column(String(32), nullable=False)
    source_id: Mapped[Optional[str]] = mapped_column(String(256), nullable=True)
    status: Mapped[str] = mapped_column(String(32), nullable=False, default="starting")
    user_prompt: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    output: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    error: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    model: Mapped[Optional[str]] = mapped_column(String(128), nullable=True)
    turn_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    created_at: Mapped[int] = mapped_column(BigInteger, nullable=False)
    started_at: Mapped[Optional[int]] = mapped_column(BigInteger, nullable=True)
    completed_at: Mapped[Optional[int]] = mapped_column(BigInteger, nullable=True)
    
    # Relationships
    events: Mapped[list["SessionEvent"]] = relationship(
        back_populates="session",
        cascade="all, delete-orphan"
    )


class SessionEvent(Base):
    """Individual events within a session."""
    __tablename__ = "session_events"
    __table_args__ = (
        Index("idx_session_events_session_ts", "session_id", "timestamp"),
    )
    
    id: Mapped[str] = mapped_column(String(64), primary_key=True)
    session_id: Mapped[str] = mapped_column(
        String(128),
        ForeignKey("sessions.id", ondelete="CASCADE"),
        nullable=False
    )
    event_type: Mapped[str] = mapped_column(String(32), nullable=False)
    timestamp: Mapped[int] = mapped_column(BigInteger, nullable=False)
    data: Mapped[str] = mapped_column(Text, nullable=False)
    
    # Relationships
    session: Mapped["Session"] = relationship(back_populates="events")

File: packages/dashboard/src/types/session.ts
export interface Session {
  id: string;
  agent_id: string;
  source: string;
  source_id?: string;
  status: 'starting' | 'running' | 'completed' | 'failed';
  user_prompt?: string;
  output?: string;
  error?: string;
  model?: string;
  turn_count: number;
  created_at: number;
  started_at?: number;
  completed_at?: number;
  key_resolution?: {
    source?: string;
    userId?: string | null;
    resolvedProviders?: string[];
    providerSources?: Record<string, { source: string; masked_key: string }>;
  } | null;
}

export interface SessionEvent {
  id: string;
  event_type: string;
  timestamp: number;
  data: Record<string, unknown>;
}

export interface SessionList {
  sessions: Session[];
  total: number;
  hasMore: boolean;
}

export interface SessionDetail extends Session {
  events: SessionEvent[];
}
